### PR TITLE
Persist state and sync between tabs

### PR DESCRIPTION
**Pull request by lukaw3d**
Part of https://github.com/oasisprotocol/oasis-wallet-web/issues/792

~Temporarily deployed to https://lukaw3d.github.io/oasis-wallet-web/ce527a5/#/~
Updated deploy https://lukaw3d.github.io/oasis-wallet-web/32fbe2b/#/

<details><summary>Old preview of <a href="https://github.com/oasisprotocol/oasis-wallet-web/compare/22369e06ec03a96cc88cd835f6a03ce98375e91e">22369e0</a></summary>

profile

</details>

### Preview of [ce527a5](https://github.com/oasisprotocol/oasis-wallet-web/compare/ce527a521a1e5bdd9c2eb9ceb6c778b2aa9d27be)

preview-ce527a5

TODO
- [x] don't store fatal errors
- [x] syncing multiple tabs and cleanup, e.g. clearAccount
    open account in two tabs http://localhost:3000/account/oasis1qzwl8nn8lt8cv4gxh85skv9rw3h0raudws66q4y5
    navigate to home on one tab (dispatches clearAccount)
    other tab loses data and shows loading
- [x] skip login:
  - [x] refactor to make networkSaga maintainable:
      always show unlock modal and
      if profile: on unlock emit setUnlockedRootState
      if profile: on skip emit event like setUnlockedRootState to use in networkSaga
      if no profile: auto click skip login emit event like setUnlockedRootState to use in networkSaga
- [x] walletId global var
- [x] close wallet = lock
- [x] change password
- [x] choose password and create profile when importing
- [x] test ui
- [x] test persistence

Deferred and moved to issue:
- auto lock timeout
- multiple profiles
- migrate old storage @metamask/browser-passworder
- test syncing tabs

### CODE REVIEW

**Review by pro-wh — COMMENTED** [review]
reading about how this works

File: src/app/state/persist/encryption.ts
Comment (lukaw3d) [reply src/app/state/persist/encryption.ts:44]: I don't understand :D 

A subset of the comment: true, we could avoid `deriveKeyFromPassword` (with a new random salt) in `decryptState` (on unlock) by returning `derivedKeyWithSalt` (with old salt) from `decryptWithPassword`. Seems safe

Comment (pro-wh) [reply src/app/state/persist/encryption.ts:44]: ya I guess, if you prefer to couple the derivation and decryption into one call

**Comment by lukaw3d** [comment]
I temporarily deployed it to https://lukaw3d.github.io/oasis-wallet-web/ce527a5/#/

**Comment by nikolaglumac** [comment]
Great work @lukaw3d 🎉 
I will do a proper review later today or tomorrow - one thing that I want to mention now (and perhaps that is something we can address in a separate PR) is that we should think about password rules. Having users set a single-character password is definitely not what we should allow.

**Comment by nikolaglumac** [comment]
Hi @lukaw3d - the flow works pretty well! I think it is a good first version - we can later iterate on it and improve it over time (e.g. by adding "remove-account" options etc).

There is this minor thing I found:
close-wallet

It should be an easy fix.

Thanks!

**Comment by lukaw3d** [comment]
Minimal edge-cases to write tests for:
- lock while viewing an account causes error with clearTransaction (fixed in https://github.com/oasisprotocol/oasis-wallet-web/pull/975/commits/966b15cfbd6da72192098c21b8cb132d5abd34cd)
- quickly unlock + lock: caused error with network fetching action (fixed in https://github.com/oasisprotocol/oasis-wallet-web/pull/975/commits/966b15cfbd6da72192098c21b8cb132d5abd34cd)
- create wallet, enable "Store private keys", type password, disable "Store private keys", next. Expect not to be stored.
- create profile, erase profile, open a new account, causes an error (fixed in https://github.com/oasisprotocol/oasis-wallet-web/pull/975/commits/787037b34b146485eb2760971c8b23c74d2bf79d)

Write test harness for interactions in multiple tabs (probably in playwright?) (defer to separate PR)

**Comment by lukaw3d** [comment]
> @nikolaglumac
> There is this minor thing I found: Once you create a profile, regardless of if you opened it later or not, "Close wallet" button is always visible

That was intentional so users can get back to login screen after clicking "Continue without the profile" (without it: user would need to create a wallet to get that same button)

**Comment by buberdds** [comment]
Loading state issue (web build only)
- open an account 
- open a new tab with the same url
- close wallet
- in the previous tab loading state "hangs"

**Comment by lukaw3d** [comment]
Fixing 1. is annoying
- I could refactor `resetRootState` to keep some slices, instead of resetting everything but: it doesn't feel correct to keep state after user clicks Close Wallet; and it probably introduces state combinations that previously weren't reachable and handled.
- Another option: I could redirect home on `resetRootState`, but only in webapp `takeLatest(resetRootState, () => { if (runtimeIs === 'webapp' && selectUnlockedStatus === 'emptyUnpersisted') location.href = '/' })`. Luckily extension with saga in background page won't support multiple tabs / popups yet. And that still needs `navigate('/'); dispatch(persistActions.lockAsync())` on click.

**Comment by csillag** [comment]
I see that it's possible to use two different accounts in two different tabs, but I can't use Mainnet on one tab and Testnet on the other. Is that an intentional choice?

File: src/config.ts
Comment (lukaw3d) [reply src/config.ts:69]: I only checked in Firefox, but I see webextension-polyfill is using the same mechanism 
https://github.com/mozilla/webextension-polyfill/blob/6e3e26c488fe6096e49e05dd7bd1e834552be869/src/browser-polyfill.js#L9-L10

File: src/app/components/Persist/UnlockForm.tsx
Comment (lukaw3d) [reply src/app/components/Persist/UnlockForm.tsx:57]: Good UX question. I've been considering the skipped profile like incognito mode, not read-only profile.
- Against:
  - We've removed search for foreign accounts, so read-only view is barely supported.
  - I was considering removing remove account ID from URL. To have fewer sources of state. That would also increase privacy - after locking wallet, someone couldn't check browser history (or click skip) to get your address.
- For:
  - Would allow wallet owner to quickly get **their account's address**. Pre-redirect URL might not be locked account (open a foreign account URL from bookmark and click skip). More complex proposal is to have a wallet pin to unlock read-only, and only ask for password before signing transactions.
  - Would allow quickly viewing **foreign account's balance** from URL/bookmark. But this is just worse search.

File: src/app/state/persist/syncTabs.ts
Comment (lukaw3d) [inline src/app/state/persist/syncTabs.ts:45]: > @csillag: I see that it's possible to use two different accounts in two different tabs, but I can't use Mainnet on one tab and Testnet on the other. Is that an intentional choice?

Yeah. I'm syncing `networkSelected` here. It would take quite some refactoring to keep it separate. I'm syncing wallet slice and it includes balances, so after switching networks it showed testnet balance on mainnet tab.

Related note in https://github.com/oasisprotocol/oasis-wallet-web/issues/792#issuecomment-1238715243
> selected network needs to be synced as long as network-specific balance is in state.wallets[*].balance

File: playwright/tests/syncTabs.spec.ts
Comment (lukaw3d) [inline playwright/tests/syncTabs.spec.ts:95]: This is failing :/ and it's a mess

- first tab initializes
- attempts to sync state from other tabs
- networkSaga initializes, dispatches 'skipUnlocking', and fetches mainnet network info
- **user switches to testnet**, networkSaga fetches testnet network info
- second tab initializes
- second tab: attempts to sync state from other tabs
- second tab: networkSaga initializes, dispatches 'skipUnlocking', and fetches mainnet network info
- first tab: sends sync state, including testnet selection
- first tab: gets synced skipUnlocking, and fetches mainnet network info
- second tab: gets synced state, including testnet selection
- first tab: dispatches selectNetwork with fetched mainnet network info
- second tab: gets synced selectNetwork with fetched mainnet network info
- both tabs are **showing mainnet** instead of testnet

Comment (lukaw3d) [reply playwright/tests/syncTabs.spec.ts:95]: https://github.com/oasisprotocol/oasis-wallet-web/pull/975/commits/c125984a415bd1a3fd9dd61c9ce6f2634e8ef8eb added a timeout to wait for tabs to sync, and only skipUnlocking if tabs don't sync

**Comment by lukaw3d** [comment]
Fixed everything, maybe. Deployed to https://lukaw3d.github.io/oasis-wallet-web/32fbe2b/#/

**Comment by buberdds** [comment]
Routing bug still occurs and it's not related to foreign accounts. Repro steps: import two accounts using mnemonic, try to open the second opened account in the second tab / fail. However, I can open the second account using Account Switcher in the second tab, but selectedWallet value will be now incorrect in the first tab, right?

**Comment by lukaw3d** [comment]
@buberdds I'm not able to reproduce this
Can you make more specific steps (when to open another tab, what URL, choosing persistence?)

I tried:
- tab1: http://localhost:3000/create-wallet, open acc1 + acc2 without persistence
- tab2: http://localhost:3000/, switch to acc2 without persistence, send transaction from acc2
- tab1: send transaction from acc1
.
- tab1: http://localhost:3000/create-wallet, open acc1 + acc2 with persistence
- tab2: http://localhost:3000/, switch to acc2, send transaction from acc2
- tab1: send transaction from acc1
.
- tab1: http://localhost:3000/create-wallet, open acc1 without persistence
- tab2: http://localhost:3000/create-wallet, open acc2 without persistence, send transaction from acc2
- tab1: send transaction from acc1
.
- tab1: http://localhost:3000/create-wallet
- tab2: http://localhost:3000/
- tab1: open acc1 without persistence
- tab2: home, create wallet, open acc2 without persistence, send transaction from acc2
- tab1: send transaction from acc1

**Comment by buberdds** [comment]
- tab1 open acc1 + acc2 via mnemonic without persistence - acc1 is open
- open tab2 - enter acc 2 url - FAIL 
As a workaround in tab2 I can open acc2 using Account Switcher, but if wallet slice is synced is value of `selectedWallet` in tab1 correct?

**Comment by lukaw3d** [comment]
Ah yeah, that is currently treated as a foreign account.

Workaround should work: wallet slice (including `selectedWallet`) is synced from tab1 to tab2 when it opens. But after initializing, `selectWallet` action isn't synced: https://github.com/oasisprotocol/oasis-wallet-web/blob/32fbe2b08ffa05a8893e03fd05f2527ba7967c03/src/app/state/persist/syncTabs.ts#L41-L49

So tab1 will continue to have selectedWallet=acc1 after tab2 switched accounts

**Review by buberdds — APPROVED** [review]
I don't have any other comments and I am fine with merging with "redirect" issue just to move forward, but would wait for @pro-wh approval related to sec.

**Comment by pro-wh** [comment]
oh uh

File: src/app/state/persist/saga.ts
Comment (pro-wh) [inline src/app/state/persist/saga.ts:152]: security review: what's this for? coverage indicates we don't test this. do we catch this somewhere?

File: src/app/state/persist/saga.ts
Comment (pro-wh) [inline src/app/state/persist/saga.ts:24]: non-security: does this result in every open tab trying to write on each change?

File: src/app/state/persist/index.ts
Comment (pro-wh) [inline src/app/state/persist/index.ts:25]: security review: only uses of `localStorage` in master branch are related to theme selection

non-security: tests use `window.localStorage`, app code uses `localStorage`, previous app code uses `window.localStorage && localStorage`

File: src/app/state/persist/saga.ts
Comment (pro-wh) [inline src/app/state/persist/saga.ts:67]: security review: callsite 1/2, stores only encrypted data :+1:

File: src/app/state/persist/saga.ts
Comment (pro-wh) [inline src/app/state/persist/saga.ts:160]: security review: callsite 2/2, stores only encrypted data :+1:

File: src/app/state/persist/encryption.ts
Comment (pro-wh) [inline src/app/state/persist/encryption.ts:48]: security review: :information_source: https://github.com/dchest/tweetnacl-js#naclsecretboxopenbox-nonce-key

**Review by pro-wh — APPROVED** [review]
:shrug: no security blockers from me

File: src/app/state/persist/saga.ts
Comment (lukaw3d) [reply src/app/state/persist/saga.ts:152]: Context: Next line skips writing on each change if wallet is locked.
For example:
- user might switch network
- it starts loading
- switch to another tab, click lock
- network info loads and dispatches networkSelected
- skip writing because we don't have the key to write
- after unlocking old network is still selected

I think it's acceptable to ignore almost all actions like this, except when silently losing an account:
- add an account from private key
- switch to another tab, click lock
- account info loads and dispatches addWallet
- throw error because we don't have the key to write

Comment (pro-wh) [reply src/app/state/persist/saga.ts:152]: couldn't understand that

is the network selection somehow related with unlocking?

in the second example, why wouldn't we have the key to write? is there a way to add an account from private key before setting up a password?

Comment (lukaw3d) [reply src/app/state/persist/saga.ts:152]: > is the network selection somehow related with unlocking?

If you mean in relation to "after unlocking old network is still selected": selected network is persisted after network info loads. So normally it would keep last selected network after unlocking.

> in the second example, why wouldn't we have the key to write? is there a way to add an account from private key before setting up a password?

when user locks one tab that syncs to other tabs and it removes stringifiedEncryptionKey from state. So we can't encrypt the new account to storage

File: src/app/components/Persist/ChoosePasswordFields.tsx
Comment (lukaw3d) [reply src/app/components/Persist/ChoosePasswordFields.tsx:34]: yep, I should also throw inside setPasswordAsync, if user somehow gets past UI

File: src/app/state/persist/saga.ts
Comment (pro-wh) [reply src/app/state/persist/saga.ts:152]: is there a more general way to stop sagas (?) when locking?

thanks for explaining

Comment (lukaw3d) [reply src/app/state/persist/saga.ts:152]: There might be, with `sagaMiddleware.run(rootSagas).cancel()`, but I haven't looked at actually implementing it and what the possible consequences could be.

None-general way would be to add `race({ task, cancel: cancel: take(persistActions.lockAsync) })` to most tasks, except the infinitely running ones. Or alternatively wait for most tasks to finish before locking (except infinitely running).

I rate it: too much effort

Comment (pro-wh) [reply src/app/state/persist/saga.ts:152]: k

File: playwright/tests/syncTabs.spec.ts
Comment (lukaw3d) [inline playwright/tests/syncTabs.spec.ts:59]: This is failing on CI, but not locally

Comment (lukaw3d) [reply playwright/tests/syncTabs.spec.ts:59]: I haven't found a reason why it matches differently on CI, only fixed it with https://github.com/oasisprotocol/oasis-wallet-web/commit/77be85730f689ec2e9586c0a4c399c2b6fbe5ce7.
`await expect(tab2.getByRole('button', { name: 'Unlock' })).toBeHidden()` on CI matched
```tsx
<a aria-label="Open wallet">
  <button tabindex="-1">
    <div>
      <svg aria-label="Unlock">..</svg>
      Open wallet
</
```

**Comment by iSQL** [comment]
Hello,

Is there any info when will this be published?

**Comment by lukaw3d** [comment]
@iSQL I'm not sure. We'll want to deploy it after we move wallet to a new domain (so there's no confusion about saved data on current domain)

and perhaps finish more of https://github.com/oasisprotocol/oasis-wallet-web/issues/792#issuecomment-1269273318

**Comment by lukaw3d** [comment]
@iSQL it is finally published

**Pull request merged**
